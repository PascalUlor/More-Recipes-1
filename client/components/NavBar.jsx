import React, { Component } from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import { Link } from 'react-router-dom';
import jwt from 'jsonwebtoken';
import GuestNavBar from './navBar/GuestNavBar.jsx';
import UserNavBar from './navBar/UserNavBar.jsx';
import logOutRequest from '../actions/actionCreators/logOutActions';
import verifyToken from '../utils/verifyToken';


/**
 * @description HOC for Navigation Bar component
 *
 * @class NavBar
 *
 * @extends Component
 */
export class NavBar extends Component {
  /**
   * @description handles decoding of token for current logged in user
   *
   * @method componentDidMount
   *
   * @returns { * } null
   */
  componentDidMount() {
    const userDecodedInfo = jwt.decode(localStorage.getItem('jwtToken'));
    if (localStorage.getItem('jwtToken') && userDecodedInfo !== null) {
      this.setState({ username: userDecodedInfo.username });
    }
  }
  /**
   * @description handles on logout event
   *
   * @return {*} null
   */
  handleLogOut() {
    this.props.logOut();
    this.context.router.history.push('/');
    window.location.reload();
  }
  /**
   * @description renders Navigation bar
   *
   * @returns { jsx } jsx - renders NavBar component
  */
  render() {
    const { isAuthenticated, user } = this.props.authenticatedUser;
    return (
      <div className="container pl-0 pr-0">
        <header className="mb-5">
          <nav className="navbar navbar-toggleable-md navbar-light fixed-top">
            <button className="navbar-toggler navbar-toggler-right"
              type="button" data-toggle="collapse" data-target="#navbarItems"
              aria-controls="navbarItems" aria-expanded="false"
              aria-label="Toggle navigation">
              <span className="navbar-toggler-icon"></span>
            </button>
            <Link className="navbar-brand" to="/">
              <img src='/images/logo.png' width="45" height="32"
                className="d-inline-block align-center" alt="More Recipe Logo"/>
              <span id="site-name">More Recipes</span>
            </Link>
            { (isAuthenticated && verifyToken()) ?
              <UserNavBar
                currentUsername={user.username}
                logOut={this.handleLogOut.bind(this)}/> :
              <GuestNavBar/>
            }
          </nav>
        </header>
      </div>
    );
  }
}

NavBar.propTypes = {
  authenticatedUser: PropTypes.shape({
    isAuthenticated: PropTypes.bool.isRequired,
    user: PropTypes.shape().isRequired
  }).isRequired,
  logOut: PropTypes.func.isRequired
};

NavBar.contextTypes = {
  router: PropTypes.shape().isRequired
};

/**
 * @description maps redux state to props
 *
 * @param { object } state - holds nav bar state
 *
 * @return { object } props - returns mapped props from state
 */
const mapStateToProps = state => ({
  authenticatedUser: state.authenticatedUser,
});
const mapDispatchToProps = dispatch => ({
  logOut: () => dispatch(logOutRequest())
});

export default connect(mapStateToProps, mapDispatchToProps)(NavBar);
