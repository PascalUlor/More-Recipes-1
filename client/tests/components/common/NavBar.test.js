/* eslint-disable */
import React from 'react';
import { Provider } from 'react-redux';
import { BrowserRouter as Router } from 'react-router-dom';
import toJson from 'enzyme-to-json';
import sinon from 'sinon';
import { NavBar } from '../../../components/NavBar.jsx';


const props = {
  authenticatedUser: {
    isAuthenticated: false,
    user: {}
  },
  logOut: jest.fn(),
  router: {
    history: {
      push: jest.fn()
    }
  }
};

const store = mockStore({});
const shallowWrapper = mount((
  <Provider store={store}>
    <Router><NavBar {...props}/></Router>
  </Provider>
));
window.localStorage = localStorage;
window.localStorage.setItem('jwtToken', 'aygdyusffgyugfufsfsffssffs');

describe('<NavBar/>', () => {
  it('should render without crashing', () => {
    expect(toJson(shallowWrapper)).toMatchSnapshot();
  });

  describe('Test for non authenticated Navbar component', () => {
    it('should render correctly for unauthenticated users', () => {
      expect(shallowWrapper.find('GuestNavBar').length).toBe(1);
    });
  });

  // describe('Test for authenticated Navbar component', () => {
  //   const newProps = {
  //     authenticatedUser: {
  //       isAuthenticated: true,
  //       user: {
  //         id: 1,
  //         username: '',
  //         email: ''
  //       }
  //     },
  //     logOut: jest.fn(),
  //     router: {
  //       history: {
  //         push: jest.fn()
  //       }
  //     }
  //   };
  //   sinon.spy(NavBar.prototype, 'handleLogOut');
  //   it('should render correctly for authenticated users', () => {
  // const verifyTokenObj = {
  //   verifyToken () => 1,
  // };
  // sinon.stub(verifyTokenObj, 'verifyToken');
  //     const shallowWrapper = mount((
  //       <Provider store={store}>
  //         <Router><NavBar {...newProps} /></Router>
  //       </Provider>
  //     ));
  //     expect(shallowWrapper.find('UserNavBar').length).toBe(1);
  //     // console.log(shallowWrapper.html());
  //   });
  // });
  // it('logOut should work correctly', () => {
  //   const logOut = wrapper.find('.btn-default');
  //   logOut.simulate('click', event);
  //   expect(Navbar.prototype.handleLogout.calledOnce).toEqual(true);
  // });
  // });
  // it('should call onChange() event', () => {
  //   mountWrapper.find('SignupForm')
  //     .find('TextFieldGroup').at(0).find('input')
  //     .simulate('change', {
  //       target: {
  //         name: 'fullName',
  //         value: 'Richie Rich',
  //       },
  //     });
  //   expect(mountWrapper.find('SignupPage')
  //     .instance().state.fullName).toEqual('Richie Rich');
  // });
});
