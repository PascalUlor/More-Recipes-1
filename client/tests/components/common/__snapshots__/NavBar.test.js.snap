// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<NavBar/> should render without crashing 1`] = `
<Provider
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <BrowserRouter>
    <Router
      history={
        Object {
          "action": "POP",
          "block": [Function],
          "createHref": [Function],
          "go": [Function],
          "goBack": [Function],
          "goForward": [Function],
          "length": 1,
          "listen": [Function],
          "location": Object {
            "hash": "",
            "pathname": "blank",
            "search": "",
            "state": undefined,
          },
          "push": [Function],
          "replace": [Function],
        }
      }
    >
      <NavBar
        authenticatedUser={
          Object {
            "isAuthenticated": false,
            "user": Object {},
          }
        }
        logOut={[MockFunction]}
        router={
          Object {
            "history": Object {
              "push": [MockFunction],
            },
          }
        }
      >
        <div
          className="container pl-0 pr-0"
        >
          <header
            className="mb-5"
          >
            <nav
              className="navbar navbar-toggleable-md navbar-light fixed-top"
            >
              <button
                aria-controls="navbarItems"
                aria-expanded="false"
                aria-label="Toggle navigation"
                className="navbar-toggler navbar-toggler-right"
                data-target="#navbarItems"
                data-toggle="collapse"
                type="button"
              >
                <span
                  className="navbar-toggler-icon"
                />
              </button>
              <Link
                className="navbar-brand"
                replace={false}
                to="/"
              >
                <a
                  className="navbar-brand"
                  href="/"
                  onClick={[Function]}
                >
                  <img
                    alt="More Recipe Logo"
                    className="d-inline-block align-center"
                    height="32"
                    src="/images/logo.png"
                    width="45"
                  />
                  <span
                    id="site-name"
                  >
                    More Recipes
                  </span>
                </a>
              </Link>
              <GuestNavBar>
                <div
                  className="collapse navbar-collapse"
                  id="navbarItems"
                >
                  <ul
                    className="navbar-nav ml-auto text-center"
                  >
                    <li
                      className="nav-item"
                    >
                      <NavLink
                        activeClassName="active"
                        activeStyle={
                          Object {
                            "backgroundColor": "whitesmoke",
                            "color": "darkcyan",
                            "fontWeight": "600",
                          }
                        }
                        ariaCurrent="true"
                        className="nav-link"
                        to="/"
                      >
                        <Route
                          path="/"
                        >
                          <Link
                            aria-current={false}
                            className="nav-link"
                            replace={false}
                            to="/"
                          >
                            <a
                              aria-current={false}
                              className="nav-link"
                              href="/"
                              onClick={[Function]}
                            >
                              Home
                            </a>
                          </Link>
                        </Route>
                      </NavLink>
                    </li>
                    <li
                      className="nav-item"
                    >
                      <NavLink
                        activeClassName="active"
                        activeStyle={
                          Object {
                            "backgroundColor": "whitesmoke",
                            "color": "darkcyan",
                            "fontWeight": "600",
                          }
                        }
                        ariaCurrent="true"
                        className="nav-link"
                        to="/recipes"
                      >
                        <Route
                          path="/recipes"
                        >
                          <Link
                            aria-current={false}
                            className="nav-link"
                            replace={false}
                            to="/recipes"
                          >
                            <a
                              aria-current={false}
                              className="nav-link"
                              href="/recipes"
                              onClick={[Function]}
                            >
                              All Recipes
                            </a>
                          </Link>
                        </Route>
                      </NavLink>
                    </li>
                    <li
                      className="nav-item"
                    >
                      <Link
                        className="nav-link"
                        replace={false}
                        to="/signin"
                      >
                        <a
                          className="nav-link"
                          href="/signin"
                          onClick={[Function]}
                        >
                          Login
                        </a>
                      </Link>
                    </li>
                    <li
                      className="nav-item"
                    >
                      <Link
                        className="nav-link nav-button"
                        replace={false}
                        to="/signup"
                      >
                        <a
                          className="nav-link nav-button"
                          href="/signup"
                          onClick={[Function]}
                        >
                          Register
                        </a>
                      </Link>
                    </li>
                  </ul>
                </div>
              </GuestNavBar>
            </nav>
          </header>
        </div>
      </NavBar>
    </Router>
  </BrowserRouter>
</Provider>
`;
